### PR TITLE
Shield unsafe journal plugins from corner case, #28541 

### PR DESCRIPTION
**Pull request by fredfp**
Fix for https://github.com/akka/akka/issues/28541

### CODE REVIEW

**Comment by raboof** [comment]
OK TO TEST

**Comment by fredfp** [comment]
Testing this is not trivial. I have a working test case but it is rather slow and ugly. I'll write another one if you agree with this fix.

**Comment by patriknw** [comment]
Could you please clarify with a code example to illustrating when it's broken?

**Comment by fredfp** [comment]
Sure, see my ugly test here: https://github.com/fredfp/akka/commit/acb1a7d55f30a0973f01ab72eb2328a0d093104d

It triggers a persistence failure with cause `akka.pattern.CircuitBreakerOpenException`.

I can also explain what happens if you need

File: akka-persistence/src/main/scala/akka/persistence/journal/AsyncWriteJournal.scala [hunk 1]
```diff
@@ -69,7 +69,8 @@ trait AsyncWriteJournal extends Actor with WriteJournalBase with AsyncRecovery {
         val atomicWriteCount = messages.count(_.isInstanceOf[AtomicWrite])
         val prepared = Try(preparePersistentBatch(messages))
         val writeResult = (prepared match {
-          case Success(prep) =>
+          case Success(prep) if prep.isEmpty => Future.successful(Nil)
+          case Success(prep)                 =>
```
Comment (patriknw) [inline akka-persistence/src/main/scala/akka/persistence/journal/AsyncWriteJournal.scala]: Let's see if I understand this correctly.

`messages: immutable.Seq[PersistentEnvelope]`, where PersistentEnvelope can be AtomicWrite or NonPersistentRepr. NonPersistentRepr are used for the defer.

`preparePersistentBatch` collects AtomicWrite, so if messages only contains NonPersistentRepr then `asyncWriteMessages` are called with empty collection.

So if empty then we continue right away instead. The ordering of the replies should be handled by `Resequencer` anyway.

**Review by patriknw — APPROVED** [review]
LGTM, good catch

File: akka-persistence/src/main/scala/akka/persistence/journal/AsyncWriteJournal.scala (hunk 1 above)
Comment (fredfp) [reply akka-persistence/src/main/scala/akka/persistence/journal/AsyncWriteJournal.scala]: Exactly

File: akka-persistence/src/main/scala/akka/persistence/journal/AsyncWriteJournal.scala [hunk 2]
```diff
@@ -69,7 +69,8 @@ trait AsyncWriteJournal extends Actor with WriteJournalBase with AsyncRecovery {
         val atomicWriteCount = messages.count(_.isInstanceOf[AtomicWrite])
         val prepared = Try(preparePersistentBatch(messages))
         val writeResult = (prepared match {
-          case Success(prep) =>
+          case Success(prep) if prep.isEmpty => Future.successful(Nil)
```
Comment (renatocaval) [inline akka-persistence/src/main/scala/akka/persistence/journal/AsyncWriteJournal.scala]: It's not obvious why we need it. For instance, what 'prepare' means here? And why we should not pass an empty `prep` to `asyncWriteMessages`? 

Can we add a block of comment explaining what is happening and why we need it?

**Review by octonato — COMMENTED** [review]
LGTM, but I think we need some more comments and explantion so we can track why we wrote the code as such.

File: akka-persistence/src/main/scala/akka/persistence/journal/AsyncWriteJournal.scala (hunk 2 above)
Comment (fredfp) [reply akka-persistence/src/main/scala/akka/persistence/journal/AsyncWriteJournal.scala]: see update and let me know if that made things clearer. Thanks!

Comment (renatocaval) [reply akka-persistence/src/main/scala/akka/persistence/journal/AsyncWriteJournal.scala]: Yes, it is. 
Especially that part: `most plugins fail calling head on empty Seq`

I don't think that was ever said in the discussion here. I was assuming it, but now it's confirmed and written down. Thanks!

**Review by octonato — APPROVED** [review]

**Comment by akilegaspi** [comment]
I know this is closed already but we're currently encountering a problem where the `asyncWriteMessages` is receiving an empty `Seq` and it's throwing an error when the cassandra plugin is trying to persist.

It's really weird.
